llarp/router/outbound_message_handler.cpp

```diff
@@ -190,34 +216,165 @@ namespace llarp
     return false;
   }
 
+  bool
+  OutboundMessageHandler::QueueOutboundMessage(const RouterID &remote,
+                                               Message &&msg,
+                                               const PathID_t &pathid)
+  {
+    MessageQueueEntry entry;
+    entry.message      = std::move(msg);
+    auto callback_copy = entry.message.second;
+    entry.router       = remote;
+    entry.pathid       = pathid;
+    if(outboundQueue.tryPushBack(std::move(entry))
+       != llarp::thread::QueueReturn::Success)
+    {
+      DoCallback(callback_copy, SendStatus::Congestion);
+    }
+
+    return true;
+  }
+
   void
-  OutboundMessageHandler::FinalizeRequest(const RouterID &router,
-                                          SendStatus status)
+  OutboundMessageHandler::ProcessOutboundQueue()
+  {
+    while(not outboundQueue.empty())
+    {
+      // TODO: can we add util::thread::Queue::front() for move semantics here?
+      MessageQueueEntry entry = outboundQueue.popFront();
+
+      auto itr_pair =
+          outboundMessageQueues.emplace(entry.pathid, MessageQueue());
+
+      if(itr_pair.second && !entry.pathid.IsZero())
+      {
+        roundRobinOrder.push(entry.pathid);
+      }
+
+      MessageQueue &path_queue = itr_pair.first->second;
+      if(path_queue.size() >= MAX_PATH_QUEUE_SIZE)
+      {
+        path_queue.pop();  // head drop
+      }
+      path_queue.push(std::move(entry));
+    }
+  }
+
+  void
+  OutboundMessageHandler::RemoveEmptyPathQueues()
+  {
+    if(not removedPaths.empty())
+    {
+      removedSomePaths = true;
+    }
+    else
+    {
+      removedSomePaths = false;
+    }
+
+    while(not removedPaths.empty())
+    {
+      auto itr = outboundMessageQueues.find(removedPaths.popFront());
+      if(itr != outboundMessageQueues.end())
+      {
+        outboundMessageQueues.erase(itr);
+      }
+    }
+  }
+
+  void
+  OutboundMessageHandler::SendRoundRobin()
+  {
+    // send non-routing messages first priority
+    while(!outboundMessageQueues[zeroID].empty())
+    {
+      MessageQueueEntry entry =
+          std::move(outboundMessageQueues[zeroID].front());
+      outboundMessageQueues[zeroID].pop();
+
+      Send(entry.router, entry.message);
+    }
+
+    size_t empty_count = 0;
+    size_t num_queues  = outboundMessageQueues.size();
+
+    if(removedSomePaths)
+    {
+      for(size_t i = 0; i < num_queues; i++)
+      {
+        PathID_t pathid = std::move(roundRobinOrder.front());
+        roundRobinOrder.pop();
+
+        if(outboundMessageQueues.find(pathid) != outboundMessageQueues.end())
+        {
+          roundRobinOrder.push(std::move(pathid));
+        }
+      }
+    }
+
+    size_t sent_count = 0;
+    while(sent_count
+          < MAX_OUTBOUND_MESSAGES_PER_TICK)  // TODO: better stop condition
+    {
+      PathID_t pathid = std::move(roundRobinOrder.front());
+      roundRobinOrder.pop();
+
+      if(outboundMessageQueues[pathid].size() > 0)
```

Comment:
This will have the side effect of creating an empty entry in `outboundMessageQueues` when one doesn't exist (and, in these cases, this conditional will still evaluate to `false`). 

~~On that note, is there any logic to clean up the queues when they're no longer used (timeout, etc.)?~~

I see that these entries are erased below, but what about the side effect mentioned here?

Reply:
if a `pathid` is in `roundRobinOrder` then it has a corresponding `outboundMessageQueues` entry.